File: PHP/tower_of_hanoi.php
<?php

/**
 * In the class problem of the towers of Hanoi, you have 3 towers and N disks of
 * different sizes which can slide onto any tower. The puzzle starts with disks
 * sorted in ascending order of size from top to bottom(ie. each disk sits on top
 * of an even bigger one). You have the following constraints:
 * 1. Only one disk can be moved at a time.
 * 2. A disk is slide off the top of one tower onto another tower.
 * 3. A disk cannot be placed on top of a smaller disk.
 * Write a program to move disks from the first tower to the last using stacks.
 */

/**
 * Time: O(2^n) where n is size of the height
 * Space: O(2^n)
 */
function towerOfHanoi($height, $src, $aux, $dst) {
    if ($height >= 1) {
        // move n-1 disk from src to aux stack
        towerOfHanoi($height-1, $src, $dst, $aux);
        // move disk n from src to dst
        $disk = array_pop($src[1]);
        array_push($dst[1], $disk);
        echo "Move disk from {$src[0]} to {$dst[0]}" . PHP_EOL;
        
        // move n-1 disk from aux stack to dst
        towerOfHanoi($height-1, $aux, $src, $dst);
    }
}

$src1 = ['src', [5,3,1]];
$aux1 = ['aux', []];
$dst1 = ['dst', []];

towerOfHanoi(3, $src1, $aux1, $dst1);

File: Python/tower_of_hanoi.py
##################################
### Title: Tower of Hanoi ########
### Author: GuoChen Hou   ########
##################################

from ADT.Stack import Stack_array

def solve_tower_recurs(number, source, temp, dest):
	"""
	Recursive solution of the Tower of Hanoi problem.

	if n > 0:
		1. move n-1 disks from source to temp using dest
		2. move 1 disk from source to dest using temp
		3. move n-1 disks from temp to dest using source
	"""
	if number > 0:
		solve_tower_recurs(number-1, source, dest, temp)
		print 'Move disk %d from peg %s to peg %s.' % (number, source, dest)
		solve_tower_recurs(number-1, temp, source, dest)

def solve_tower(number):
	"""
	Solve tower of hanoi problem using stack ADT and recursive solution.

	Initialize 3 stacks to with each stack symbolise each of the peg.
	"""
	towers = []
	# initialize a list reference each of the 3 towers stacks
	for i in range(number):
		stack = Stack_array()
		towers.append(stack)
	# put all pegs to starting tower in ascending order from top to bottom
	for i in range(number-1, -1, -1):
		towers[0].push(i)
	# recursive solution
	solve_tower_stack(number, towers[0], towers[1], towers[2])

def solve_tower_stack(number, source, temp, dest):
	"""
	Solve tower of hanoi problem using stack ADT and recursive solution.
	"""
	if number > 0:
		solve_tower_stack(number-1, source, dest, temp)
		move = source.pop()
		dest.push(move)
		solve_tower_stack(number-1, temp, source, dest)

if __name__ == "__main__":
	num = int(raw_input('Number of disks: '))

	solve_tower_recurs(num, 'A', 'B', 'C')
	solve_tower(num)